### PR TITLE
Lower bound checks for GeneralizedTime and UTCTime

### PR DESCRIPTION
**Pull request by job**
[ITU-T X.690 / ISO/IEC 8825-1](https://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf) section 11.7 and section 11.8 impose specific constraints on how _GeneralizedTime_ and _UTCTime_ can be encoded in BER/CER/DER. Following from these constraints a minimum length can be derived.

ITU-T X.690 Section 11.7.2 and 11.8.2 specify for _GeneralizedTime_ and _UTCTime_: "The seconds element shall always be present."

Checking the length in this context can potentially help prevent applications from interpreting an invalid GeneralizedTime as a valid UTCTime.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by job** [comment]
Seems the fuzzer test failed for unrelated reasons:

```
2024-02-05T12:03:50.7581639Z /usr/bin/ld: final link failed: No space left on device
2024-02-05T12:03:50.7592349Z /usr/bin/ld: final link failed: No space left on device
2024-02-05T12:03:50.7614023Z /usr/bin/ld: final link failed: No space left on device
2024-02-05T12:03:50.7810353Z clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
2024-02-05T12:03:50.7819692Z clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
2024-02-05T12:03:50.7838002Z make[1]: *** [Makefile:18328: test/ciphername_test] Error 1
2024-02-05T12:03:50.7838788Z make[1]: *** Waiting for unfinished jobs....
2024-02-05T12:03:50.7851387Z clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
2024-02-05T12:03:50.7865396Z make[1]: *** [Makefile:18357: test/cmactest] Error 1
2024-02-05T12:03:50.7873862Z make[1]: *** [Makefile:18343: test/clienthellotest] Error 1
2024-02-05T12:03:51.0431217Z make[1]: Leaving directory '/src/openssl32'
2024-02-05T12:03:51.0444015Z make: *** [Makefile:2916: build_sw] Error 2
2024-02-05T12:03:55.3438227Z 2024-02-05 12:03:55,343 - root - ERROR - Building fuzzers failed.
```

File: crypto/asn1/tasn_dec.c [hunk 1]
```diff
@@ -921,6 +921,14 @@ static int asn1_ex_c2i(ASN1_VALUE **pval, const unsigned char *cont, int len,
             ERR_raise(ERR_LIB_ASN1, ASN1_R_UNIVERSALSTRING_IS_WRONG_LENGTH);
             goto err;
         }
+        if (utype == V_ASN1_GENERALIZEDTIME && (len < 15)) {
+            ERR_raise(ERR_LIB_ASN1, ASN1_R_GENERALIZEDTIME_IS_TOO_SHORT);
```
Comment (bernd-edlinger) [inline crypto/asn1/tasn_dec.c:925]: This check is more strict than what we have here:
https://github.com/openssl/openssl/blob/22f82d457c06289ec66a627a3d11649d83beff88/crypto/asn1/a_time.c#L102-L110
so I think, it is likely to cause issues.

Comment (job) [reply crypto/asn1/tasn_dec.c:925]: Can you substantiate what issues you expect to be caused? I'm skeptical implementers would consciously rely on invalid encodings, X.690 Section 11.7 reads unambiguous to me.

I'm not arguing for full conformance to RFC 5280 (although such strictness seems to work well in other crypto libraries), but to ensure that at the very least an application can't mistakenly interpret an invalid _GeneralizedTime_ as a valid _UTCTime_ (for example, if the decision tree is based on the length).

The code path followed in `a_time.c` when `strict` is set to `1` arguably is far stricter than what I propose here.

**Review by t8m — APPROVED** [review]

File: test/asn1_decode_test.c [hunk 2]
```diff
@@ -161,6 +162,54 @@ static int test_uint64(void)
...
+    };
+    const unsigned char *p;
+    int der_len;
+    ASN1_GENERALIZEDTIME *gentime;
+
+    p = der;
+    der_len = sizeof(der);
+    gentime = d2i_ASN1_GENERALIZEDTIME(NULL, &p, der_len);
+
+    if (gentime != NULL)
```
Comment (t8m) [inline test/asn1_decode_test.c:182]: Please use `if (!TEST_ptr(gentime))` instead. Also please release the returned object.

File: test/asn1_decode_test.c [hunk 3]
```diff
@@ -161,6 +162,54 @@ static int test_uint64(void)
...
+    };
+    const unsigned char *p;
+    int der_len;
+    ASN1_UTCTIME *utctime;
+
+    p = der;
+    der_len = sizeof(der);
+    utctime = d2i_ASN1_UTCTIME(NULL, &p, der_len);
+
+    if (utctime != NULL)
```
Comment (t8m) [inline test/asn1_decode_test.c:205]: Same as above

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

File: crypto/asn1/tasn_dec.c (hunk 1 above)
Comment (bernd-edlinger) [reply crypto/asn1/tasn_dec.c:925]: I dont know why this check is more permissive, but your test vectors are rejected by ASN1_TIME_check,
but slightly modified test vectors i.e. without seconds pass the check in ASN1_TIME_check, as this experiment
shows:
```
diff --git a/test/asn1_decode_test.c b/test/asn1_decode_test.c
index f112dd7034..e33f946520 100644
--- a/test/asn1_decode_test.c
+++ b/test/asn1_decode_test.c
@@ -166,10 +166,10 @@ static int test_uint64(void)
 
 static int test_gentime(void)
 {
-    /* Underflowing GeneralizedTime 161208193400Z (YYMMDDHHMMSSZ) */
+    /* Underflowing GeneralizedTime 209912312359Z (YYYYMMDDHHMMZ) */
     const unsigned char der[] = {
-        0x18, 0x0d, 0x31, 0x36, 0x31, 0x32, 0x30, 0x38, 0x31,
-        0x39, 0x33, 0x34, 0x30, 0x30, 0x5a,
+        0x18, 0x0d, 0x32, 0x30, 0x39, 0x39, 0x31, 0x32, 0x33,
+        0x31, 0x32, 0x33, 0x35, 0x39, 0x5a,
     };
     const unsigned char *p;
     int der_len, rc = 1;
@@ -179,7 +179,8 @@ static int test_gentime(void)
     der_len = sizeof(der);
     gentime = d2i_ASN1_GENERALIZEDTIME(NULL, &p, der_len);
 
-    if (!TEST_ptr_null(gentime))
+    if (!TEST_true(gentime != NULL
+            && ASN1_GENERALIZEDTIME_check(gentime)))
         rc = 0; /* fail */
 
     ASN1_GENERALIZEDTIME_free(gentime);
@@ -190,10 +191,10 @@ static int test_gentime(void)
 
 static int test_utctime(void)
 {
-    /* Underflowing UTCTime 0205104700Z (MMDDHHMMSSZ) */
+    /* Underflowing UTCTime 9912312359Z (YYMMDDHHMMZ) */
     const unsigned char der[] = {
-        0x17, 0x0b, 0x30, 0x32, 0x30, 0x35, 0x31, 0x30,
-        0x34, 0x37, 0x30, 0x30, 0x5a,
+        0x17, 0x0b, 0x39, 0x39, 0x31, 0x32, 0x33, 0x31,
+        0x32, 0x33, 0x35, 0x39, 0x5a,
     };
     const unsigned char *p;
     int der_len, rc = 1;
@@ -203,7 +204,8 @@ static int test_utctime(void)
     der_len = sizeof(der);
     utctime = d2i_ASN1_UTCTIME(NULL, &p, der_len);
 
-    if (!TEST_ptr_null(utctime))
+    if (!TEST_true(utctime != NULL
+            && ASN1_UTCTIME_check(utctime)))
         rc = 0; /* fail */
 
     ASN1_UTCTIME_free(utctime);

```

Comment (bernd-edlinger) [reply crypto/asn1/tasn_dec.c:925]: so IMHO ASN1_XXXTIME_check should also be changed to not allow this any more,
and of course a CHANGES entry should be added documenting this change.

Comment (job) [reply crypto/asn1/tasn_dec.c:925]: Yes, `ASN1_TIME_check()` ultimately wraps around `ossl_asn1_utctime_to_tm()` which indeed does a few useful sanity checks, but on the `d2i_ASN1_GENERALIZEDTIME() / d2i_ASN1_UTCTIME()` paths, `ossl_asn1_utctime_to_tm()` isn't invoked.

Comment (job) [reply crypto/asn1/tasn_dec.c:925]: @bernd-edlinger I updated `ASN1_XXXTIME_check()` too

**Comment by t8m** [comment]
**OTC: Is it OK to enforce the requirement of the standard to have the seconds value always present?**

**Comment by job** [comment]
> **OTC: Is it OK to enforce the requirement of the standard to have the seconds value always present?**

I suspect that enforcing the requirement will remove a potential for malleability.

**Comment by t8m** [comment]
The concern here is with breaking 3rd party applications that (albeit incorrectly) use the GeneralizedTime or UTCTime without the seconds part. We are thinking about doing this in 4.0.

**Comment by t8m** [comment]
**OTC: In 3.x we could accept a documentation change to say that these functions accept incorrectly encoded time values without the seconds part, but we may consider removing this handling by 4.0 release.**

**Comment by job** [comment]
> **OTC: In 3.x we could accept a documentation change to say that these functions accept incorrectly encoded time values without the seconds part, but we may consider removing this handling by 4.0 release.**

Dear OTC,

I'm a bit at a loss here. As it stands, the OpenSSL X.690 implementation can _accept_ and _produce_ invalid GeneralizedTime and UTCTime encodings (as the test vectors demonstrate).

The objective of the X.690 standard was to specify an unambiguous way to foster interoperability across different encoder/decoder implementations. A secondary advantage of following the rules is that a GeneralizedTime and UTCTime are visually distinct.

The OTC appears to reason that maintaining bug-for-bug compatibility with older versions of OpenSSL is important. But what about being compatible with other X.690 implementations which _do_ follow the specification?

If the concern is that the proposed bugfix breaks API-contracts, I beg to differ: it actually restores an API contract promise. An unassuming user might believe that the OpenSSL X.690 produces canonical forms, which in fact it doesn't.

I fail to see what is gained by putting the fix off until 4.0.0 or later. Reading the Release policy this seems appropriate for a minor release.

**Comment by t8m** [comment]
I removed the 4.0.0 milestone so we will revisit this again next week.

**Comment by t8m** [comment]
**OTC: We reconsidered our decision after @job 's comment and we are OK with accepting this to be merged to the master branch once reviewed and approved.**

**Comment by job** [comment]
rebased

**Review by t8m — APPROVED** [review]

**Review by bernd-edlinger — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Comment by rhenium** [comment]
> [ITU-T X.690 / ISO/IEC 8825-1](https://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf) section 11.7 and section 11.8 impose specific constraints on how _GeneralizedTime_ and _UTCTime_ can be encoded in BER/CER/DER. Following from these constraints a minimum length can be derived.
> 
> ITU-T X.690 Section 11.7.2 and 11.8.2 specify for _GeneralizedTime_ and _UTCTime_: "The seconds element shall always be present."

Please note that X.690's chapter 11 applies to CER and DER, but not to BER. I'm not sure what exactly is the API contract here. Are d2i_*() functions supposed to accept DER only?

If it's the case, I'd suggest the constraint in X.690 section 11.7.1 and 11.8.1 ("The encoding shall terminate with a Z") be applied for consistency. That would make it reject these: https://github.com/openssl/openssl/blob/eadd8c4727b703049e4d2764751cb04f3108434d/test/asn1_time_test.c#L119-L128

FWIW, LibreSSL 3.6.0 (2022-10-06) applies RFC 5280's rules on ASN1_UTCTIME and ASN1_GENERALIZEDTIME (i.e., no fractional seconds support on top of DER): https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.6.0-relnotes.txt

**Comment by junaruga** [comment]
> Merged to the master branch. Thank you for your contribution.

For someone who is interested in the actual commit on the master branch, it is eadd8c4727b703049e4d2764751cb04f3108434d .

**Pull request closed without merging**